Remove the ultimate ancestor and split the grand MRCA

hyanwong opened the pull request:
In a [few](https://github.com/tskit-dev/tsinfer/issues/649#issuecomment-1143628417) [issues](https://github.com/tskit-dev/tsinfer/issues/275#issuecomment-1065012704) we decided to remove the "ultimate" ancestors (the hack that means that the ancestors IDs in the ancestors file and the tree sequence align. This is needed as a precursor to #673.

This does this, removing the assert, but on my laptop I'm finding this now fails in unexpected ways in find_path etc, which makes me think that we are removing one of the ancestors required for LS matching.

hyanwong commented:
Note that the ancestors IDs and the nodes in the ancestors tree sequence won't align anyway if we allow path compression during match_ancestors.

hyanwong commented:
Actually the problem is [here](https://github.com/tskit-dev/tsinfer/blob/87f7b54c4ac6294af46cf37095f6881d1df59995/tsinfer/algorithm.py#L873). I think the issue is that at the matching algorithm assumes that all the nodes against which we match have an edge above it, we don't have this if we remove the ultimate ancestor.

jeromekelleher commented:
Yes, it's not straightforward - it's not obvious how to fix easily.

jeromekelleher commented:
We also make a hard assumption that zero if the ultimate root, which makes it very tricky to use arbitrary tree sequences as input. We'd need to fix that too. My best guess on the simplest way to fix this is to make an "virtual root" node at the end of the node table, and to add in edges from all the actual roots to this, and then change the places where we're assuming 0 to this node. But it's highly optimised code and the logic is very tricky (it's mainly about nodes that fall out of the tree, and keeping track of that) so it's not an easy thing to do.

hyanwong commented:
So are we still of the opinion that this is the best thing to do, or should we go to the fallback of deleting the ultimate ancestor from the resulting tree sequence?

If we leave it in during ancestor generation, we could set the time of the ultimate ancestor in tsinfer to `np.inf`, then it can act as a sort of sentinel, to ensure that we can't put it into a tree sequence (we can remove it before calling `tables.tree_sequence()`?

jeromekelleher commented:
I don't know to be honest. It's not really that much of a problem and we have things that urgently need to be done, so I'd rather park it.

hyanwong commented:
It's stopping me addressing #673 which does actually make a difference. So I'm happy to take an executive decision and leave it in with `np.inf` as the time (or at least, test whether this works)

hyanwong commented:
So I think I've figured out the easiest way to do this: we require the ultimate ancestor when matching. But we don't do matching on a simplified tree sequence. So the easiest thing is to only remove the ultimate ancestor when we simplify, during the final output stage. The nicest way to do this, IMO, is to remove the (single) edge connecting the ultimate ancestor (ancestor ID 0) to the grand MRCA (ancestor ID 1), then let `simplify` take care of the removal of the node.

I've adjusted this PR to do that. The only reasons it is now failing is the test where we compare the simplified output to the output when running the standard `.simplify(keep_unary=True)`, as these outputs now differ because of the removal of the ultimate ancestor. I guess that's fine, and we can adjust the tests.

Comments as to whether this is the best approach are very welcome.

hyanwong commented:
I have this now in satisfactory shape, so it's ready for review, hurrah.

hyanwong commented:
This will also fix https://github.com/tskit-dev/tsinfer/issues/622

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1924,6 +1912,156 @@ def get_mutations(self, node_id):
+            )
+            mutation_id, mutation_pos = next(mut_iter, (None, ts.sequence_length))
+        left = right
+    return True
```
What's the return for?

hyanwong replied:
Removed anyway with new code arrangement

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1924,6 +1912,156 @@ def get_mutations(self, node_id):
+    return tables.tree_sequence()
+
+
+def detach_node_over_grand_root(tables):
```
Split this into a function ``has_single_edge_over_grand_root`` . The actual taking off is trivial and can be done like:
```
if has_single_edge_over_grand_root(tables):
    tables.edges.truncate(tables.edges.num_rows - 1)
```

hyanwong replied:
Done, although we need to move the mutations afterwards too, of course

jeromekelleher reviewed (commented):
Looks good generally. Needs to be decomposed into simpler functions that we can hopefully test well in isolation.

hyanwong commented:
> Looks good generally. Needs to be decomposed into simpler functions that we can hopefully test well in isolation.

Thanks *a lot* for looking at this. I restructured the code to avoid all that sorting and turning backwards and forwards into tables/tree sequences, although that's unavoidable to loop through the trees after detaching the "virtual root-like" ancestor. I guess one or two more tests of the `has_single_edge_over_grand_root` and `has_same_root_everywhere` helper functions (should these be prefixed with an underscore?) might be good, but is this new approach OK?

Comments on algorithms etc in replied to the feedback above.

hyanwong commented on tsinfer/inference.py:
```diff
@@ -1924,6 +1912,156 @@ def get_mutations(self, node_id):
+    genomewide_mrca_id = roots.pop()
+    assert genomewide_mrca_id == tables.edges[tables.edges.num_rows - 1].parent
+    genomewide_mrca = tables.nodes[genomewide_mrca_id]
+    edges_to_remove = tables.edges.parent == genomewide_mrca_id
```
It's neater to reverse the `ts.edges()` iterator, I think, so that we can use the IDs (and if we do it on the ts rather than the tables, we are assured that the order is correct):

```python
    root_breaks = set()
    keep_edges = 0
    for e in reversed(ts.edges()):
        # all the edges at the end are the ones to the genomewide_mrca
        if e.parent == genomewide_mrca_id:
            root_breaks |= {e.left, e.right}
        else:
            keep_edges = e.id + 1
            break
    tables.edges.truncate(keep_edges)
```

jeromekelleher replied:
I'd find the while-loop version easier (and shorter) to read tbh, but :shrug:

hyanwong replied:
There's a bug in that one: I think we want to `truncate(j+1)`, which was another reason for spelling out `keep_edges` explicitly. But happy to switch back.

hyanwong commented:
NB currently failing because I've started to use the negative node indexing stuff in the upcoming `tskit 0.5.3`. We can wait to merge until that's released, perhaps (as long as the approach is correct)

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1907,6 +1909,149 @@ def get_mutations(self, node_id):
+    if ts.num_edges < 2:
+        # must have edge to grand root and above grand root
+        return False
+    last_edge = ts.edge(-1)
```
I'd find it easier to follow the logic here (and it would be easier to test) if the full-span requirement was tested first. This clause will also catch split edges like (0, x, u, vr), (x, L, u, vr).

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1907,6 +1909,149 @@ def get_mutations(self, node_id):
+        logger.info("Removing the oldest edge to detach the virtual-root-like ancestor")
+        last_edge = ts.edge(
+            -1
+        )  # Edge with oldest parent must be last in the edge table
```
I would either remove this comment or put it on the line above so we don't get this line-splitting from Black.

jeromekelleher reviewed (commented):
I've had a good look through and I *think* the logic is correct. Needs some tests for the helper functions, (i.e. class TestHasSameRootEverywhere ... try on some different wacky cases)

jeromekelleher commented:
LMK when this is ready for final review please @hyanwong

jeromekelleher commented on tests/test_inference.py:
```diff
@@ -2508,6 +2518,228 @@ def test_many_trees(self):
+    yield tables.tree_sequence(), "Multi_root_edge"
+
+
+def ts_root_test_params():
```
Nice!

jeromekelleher commented on tests/test_inference.py:
```diff
@@ -2508,6 +2518,228 @@ def test_many_trees(self):
+    @pytest.mark.parametrize("ts", ts_root_test_params())
+    def test_has_single_edge_over_grand_root(self, ts, request):
+        if request.node.callspec.id == "Single_root_edge":
+            assert tsinfer.has_single_edge_over_grand_root(ts)
```
This is brittle and needs the reader to understand the innards of pytest. Better to break up the list of examples (or to make this property a parameter of ``ts_root_test_params()``?)

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1912,6 +1908,146 @@ def get_mutations(self, node_id):
+    method with ``keep_unary`` set to True but ``filter_sites``, ``filter_populations``
+    and ``filter_individuals`` set to False).
+
+    :param bool split_mrca: If the oldest node is only a parent to a single
```
Move the bulk of this discussion to a paragraph rather than within the ``split_mrca`` parameter description - these should be short and focus on things like types, etc.

jeromekelleher reviewed (commented):
LGTM. Needs another pass to make sure we're clear on the compatibility semantics going forward, and we deal with the single-tree case nicely.

hyanwong commented:
> LGTM. Needs another pass to make sure we're clear on the compatibility semantics going forward, and we deal with the single-tree case nicely.

Great, thanks a lot for looking over this: a few people will benefit from this new feature (I mentioned it to at least 2 people at the ARG symposium). All your comments seems good improvements. I'll close them off one by one now.

hyanwong commented on tsinfer/inference.py:
```diff
@@ -1912,6 +1908,146 @@ def get_mutations(self, node_id):
+    # split it up into a set of contemporaneous nodes whenever the node children change
+
+    ts = tables.tree_sequence()
+    if not has_same_root_everywhere(ts):
```
We haven't done that though, have we? I can only find `has_same_root_everywhere` being called once in the entire file. Above we check whether there is a "virtual-root-like" ancestor, but that's not the same. For instance, this would fail if there were isolated nodes in the TS.

jeromekelleher replied:
Right, sorry, I mixed this up with has_single_edge_over_grand_root

jeromekelleher replied:
Needs a test though, currently not covered.

hyanwong replied:
I think it's in `TestRootCheckingFunctions::test_has_same_root_everywhere`?

hyanwong commented on tsinfer/inference.py:
```diff
@@ -221,22 +221,24 @@ def infer(
     engine=constants.C_ENGINE,
     progress_monitor=None,
     time_units=None,
+    simplify=True,  # Deprecated
```
Yes, this all seems good to me apart from:

> If the user explicitly sets simplify=True, we do the same thing as before (we don't do the new things - which we can, right?)

There's no path to do that right now, but I can add it.

jeromekelleher replied:
Seems worthwhile - no point in breaking people's code for no reason.

hyanwong replied:
Yep, I hadn't thought of this avenue. Added in my local changes. Will push in a tick.

hyanwong commented on tests/test_inference.py:
```diff
@@ -2508,6 +2518,228 @@ def test_many_trees(self):
+            )
+            inferred_individual = ts.individual(ts.node(inferred_node_id).individual)
+            assert sd_individual.metadata["id"] == inferred_individual.metadata["id"]
+
```
> We need a test here for the case where we _infer_ a single tree (recombination rate = ~0), verifying that we don't change the input at all. 

Sorry, I'm being dense and don't quite understand your comments here. You mean that we don't change the genotypes, or that we infer exactly the same tree as we used to generate the variant data. If you mean the same tree, then we do need to drop the oldest edge (as that won't be in the original, simulated tree)

hyanwong replied:
Thanks for clarifying just now @jeromekelleher  - I've added a check so that we simply return without doing anything if there is only a single edge to the grand MRCA, so we don't delete and recreate a node.

hyanwong commented:
Apart from addressing you comments above, I'm adding one extra thing, which is that I now call `logger.warning` when we post process within the inference step (i.e. in match samples) and we can't find the virtual-root-like-ancestor and/or the grand MRCA. I've done this because at the moment, we simply invisibly avoid splitting etc if the inferred tree sequence is not in the expected format, and I'dm prefer to know that there's something unexpected about the format of the output inferred TS in this case.

jeromekelleher commented:
SGTM. Might also be worth issuing a warning when users specify ``simplify=True`` suggesting they change over to ``post_process=True``.

hyanwong commented:
I hope I've addressed all the comments now: I've done them as a separate commit so the changes are obvious. Let me know if they look OK, @jeromekelleher, and I can squash the latest commit into the previous one.

jeromekelleher commented on CHANGELOG.rst:
```diff
@@ -13,6 +13,14 @@
+- The ``simplify`` parameter is now deprecated in favour of ``post_process``, which
+  prior to simplification, removes the "virtual-root-like" ancestor (inserted by
+  tsinfer to aid the matching process) then splits the grand MRCA into separate pieces.
+  If splitting is not required, the ``post_process` step can also be called as a
```
```suggestion
  If splitting is not required, the ``post_process`` step can also be called as a
```

jeromekelleher commented on requirements/CI-tests-complete/requirements.txt:
```diff
@@ -16,5 +16,5 @@ pytest-xdist==2.5.0
 sortedcontainers==2.4.0
 tqdm==4.64.0
-tskit==0.5.0
+tskit==0.5.3
```
Also need to update our installation requirements, probably in setup.cfg.

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -589,25 +592,41 @@ def match_samples(
+            post_process = True
+    else:
+        if post_process is not None:
+            raise ValueError("Can't specify both `simplify` and `post_process`")
```
Needs a test

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -1750,6 +1759,8 @@ def get_samples_tree_sequence(self, map_additional_sites=True):
             # root.
             assert left.shape[0] == 0
             root = tables.nodes.add_row(flags=0, time=tables.nodes.time.max() + 1)
+            ultimate = tables.nodes.add_row(flags=0, time=tables.nodes.time.max() + 2)
```
No point in running max on tables.nodes.time twice

jeromekelleher commented on tsinfer/inference.py:
```diff
@@ -640,7 +659,7 @@ def match_samples(
 
     manager.match_samples(sample_indexes, sample_times)
-    ts = manager.finalise(simplify=simplify)
+    ts = manager.finalise(do_post_process=post_process, simplify_only=simplify_only)
```
I'm not sure there's much point in sending these parameters to the finalise method - could we just do
```
if post_process:
    ts = post_process(...)
return ts
```

hyanwong replied:
Oh yes, good to have the extra eyes on this. Although it's a nice idea, it does create a variable name conflict between the `post_process` parameter and the identically named function. Do we want to rename anything to get around this?

hyanwong replied:
I can work around it by defining `_post_process` as an internal alias for the `post_process()` function, but it's a bit messy.

jeromekelleher reviewed (commented):
LGTM, spotted a few final things.

Need to be careful with our updated requirement on tskit getting carried through properly to the final packages.

jeromekelleher reviewed (approved):
LGTM let's merge. Should probably wait for the JupyterBook one to land first though

jeromekelleher commented:
Some conflicts to fix @hyanwong , we can auto merge then

hyanwong commented:
> Some conflicts to fix @hyanwong , we can auto merge then

Oh yes, sorry - didn't see this as I was checking on mobile. Should be OK now. How do I auto merge?
